Add new use cases to ML Inference Search Response Processor (#8639)
opensearch-project/documentation-website: _search-plugins/search-pipelines/ml-inference-search-response.md

    ]
  }
}
```

### Example: Externally hosted model

> This example demonstrates configuring an `ml_inference` search response processor to work with an externally hosted large language model (LLM) and map the model's response to the search extension object. Using the `ml_inference` processor, you can enable an LLM to summarize search results directly within the response. The summary is included in the `ext` field of the search response, providing seamless access to AI-generated insights alongside the original search results.

Reviewer: ```suggestion
This example demonstrates how to configure an `ml_inference` search response processor to work with an externally hosted large language model (LLM) and map the model's response to the search extension object. Using the `ml_inference` processor, you can enable an LLM to summarize search results directly within the response. The summary is included in the `ext` field of the search response, providing seamless access to AI-generated insights alongside the original search results.
```

```json
PUT /_search/pipeline/my_pipeline_request_review_llm
{
  "response_processors": [
    {
      "ml_inference": {
        "tag": "ml_inference",
        "description": "This processor is going to run llm",
        "model_id": "EOF6wJIBtDGAJRTD4kNg",
        "function_name": "REMOTE",
        "input_map": [
          {
            "context": "review"
          }
        ],
        "output_map": [
          {
            "ext.ml_inference.llm_response": "response"
          }
        ],
        "model_config": {
          "prompt": "\n\nHuman: You are a professional data analysist. You will always answer question: Which month had the lowest customer acquisition cost per new customer? based on the given context first. If the answer is not directly shown in the context, you will analyze the data and find the answer. If you don't know the answer, just say I don't know. Context: ${parameters.context.toString()}. \n\n Assistant:"
        },
        "ignore_missing": false,
        "ignore_failure": false
      }
    }
  ]
}
```
{% include copy-curl.html %}

> In this configuration, you're providing the following parameters:

Reviewer: ```suggestion
In this configuration, you've provided the following parameters:
```

> - The `model_id` specifies the ID of the generative AI model.

Reviewer: ```suggestion
- The `model_id` parameter specifies the ID of the generative AI model.
```

> - The `function_name` is set to `REMOTE`, indicating an externally hosted model.

Reviewer: ```suggestion
- The `function_name` parameter is set to `REMOTE`, indicating that the model is hosted externally.
```

> - The `input_map` maps the review field from the document to the context field expected by the model.

Reviewer: ```suggestion
- The `input_map` parameter maps the review field from the document to the context field expected by the model.
```

> - The `output_map` specifies that the model's response should be stored in `ext.ml_inference.llm_response` in the search response.

Reviewer: ```suggestion
- The `output_map` parameter specifies that the model's response should be stored in `ext.ml_inference.llm_response` in the search response.
```

> - The `model_config` includes a prompt that instructs the model how to process the input and generate a summary.

Reviewer: ```suggestion
- The `model_config` parameter includes a prompt that tells the model how to process the input and generate a summary.
```

        }
      },
      {
        "_index": "review_string_index",
        "_id": "3",
        "_score": 1,
        "_source": {
          "review": "Customer acquisition cost: March: $40, New customers: March: 750."
        }
      }
    ]
  },
  "ext": {
    "ml_inference": {
      "llm_response": """ Based on the context provided:

      - Customer acquisition cost in January was $50 and new customers were 500. So the cost per new customer was $50/500 = $0.10

      - Customer acquisition cost in February was $45 and new customers were 600. So the cost per new customer was $45/600 = $0.075

      - Customer acquisition cost in March was $40 and new customers were 750. So the cost per new customer was $40/750 = $0.053

      Therefore, the month with the lowest customer acquisition cost per new customer was March, as it had the lowest cost per customer of $0.053."""
    }
  }
}
```

### Example: Reranking search results using a text similarity model

The following example shows you how to configure an `ml_inference` search response processor with a text similarity model.

**Prerequisite**

> You must configure an externally hosted LLM for this use case. For more information about externally hosted models, see [Connecting to externally hosted models]({{site.url}}{{site.baseurl}}/ml-commons-plugin/remote-models/index/). Once you register the LLM, you can use the following request to test it. This request requires providing a `text` and `text_pair` fields within the `inputs` field:

Reviewer: ```suggestion
You must configure an externally hosted LLM for this use case. For more information about externally hosted models, see [Connecting to externally hosted models]({{site.url}}{{site.baseurl}}/ml-commons-plugin/remote-models/index/). Once you register the LLM, you can use the following request to test it. This request requires that you provide the `text` and `text_pair` fields within the `inputs` field:
```

```json
POST /_plugins/_ml/models/Ialx65IBAs32TwoK1lXf/_predict
{
  "parameters": {
    "inputs":
    {
      "text": "I like you",
      "text_pair": "I hate you"
    }
  }
}
```
{% include copy-curl.html %}

The model returns similarity scores for each input document:

```json
{
  "inference_results": [
    {
      "output": [
        {
          "name": "response",
          "dataAsMap": {
            "label": "LABEL_0",
            "score": 0.022704314440488815
          }
        }
      ],
      "status_code": 200
    }
  ]
}
```
{% include copy-curl.html %}

**Step 1: Index sample documents**

> Create an index and add some sample documents to it:

Reviewer: ```suggestion
Create an index and add some sample documents:
```

```json
PUT /_search/pipeline/my_rerank_pipeline
{
  "response_processors": [
    {
      "ml_inference": {
        "tag": "ml_inference",
        "description": "This processor runs ml inference during search response",
        "model_id": "Ialx65IBAs32TwoK1lXf",
        "model_input":"""{"parameters":{"inputs":{"text":"${input_map.text}","text_pair":"${input_map.text_pair}"}}}""",
        "function_name": "REMOTE",
        "input_map": [
          {
            "text": "diary",
            "text_pair":"$._request.query.term.diary.value"
          }
        ],
        "output_map": [
          {
            "rank_score": "$.score"
          }
        ],
        "full_response_path": false,
        "model_config": {},
        "ignore_missing": false,
        "ignore_failure": false,
        "one_to_one": true
        },
        "rerank": {
          "by_field": {
            "target_field": "rank_score",
            "remove_target_field": true
          }
        }
    }
  ]
}
```
{% include copy-curl.html %}

> In this configuration, you're providing the following parameters:

Reviewer: ```suggestion
In this configuration, you've provided the following parameters:
```

> - The `model_id` specifies the unique identifier of the text similarity model.

Reviewer: ```suggestion
- The `model_id` parameter specifies the unique identifier of the text similarity model.
```

> - The `function_name` is set to `REMOTE`, indicating that the model is hosted externally.

Reviewer: ```suggestion
- The `function_name` parameter is set to `REMOTE`, indicating that the model is hosted externally.
```

> - The `input_map` maps the `diary` field from each document to the `text` input of the model, and the search query term to the `text_pair` input.

Reviewer: ```suggestion
- The `input_map` parameter maps the `diary` field from each document to the `text` input of the model as well as the search query term to the `text_pair` input.
```

> - The `output_map`  maps the model's score to a field named `rank_score` in each document.

Reviewer: ```suggestion
- The `output_map` parameter maps the model's score to a field named `rank_score` in each document.
```

> - The `model_input` formats the input for the model, ensuring it matches the structure expected by the Predict API.

Reviewer: ```suggestion
- The `model_input` parameter formats the input for the model, ensuring that it matches the structure expected by the Predict API.
```

> - The `one_to_one` parameter is set to `true`, ensuring that the model processes each document individually, rather than batching multiple documents together.

Reviewer: ```suggestion
- The `one_to_one` parameter is set to `true`, ensuring that the model processes each document individually rather than batching multiple documents together.
```

- The `ignore_missing` parameter is set to `false`, causing the processor to fail if the mapped fields are missing from a document.
- The `ignore_failure` parameter is set to `false`, causing the entire pipeline to fail if the ML inference processor encounters an error.

> The `rerank` processor is applied after the ML inference. It reorders the documents based on the `rank_score` field generated by the ML model and then removes this field from the final results.

Reviewer: ```suggestion
The `rerank` processor is applied after ML inference. It reorders the documents based on the `rank_score` field generated by the ML model and then removes this field from the final results.
```

**Step 3: Run the pipeline**

> Now, perform a search using the created pipeline:

Reviewer: ```suggestion
Now perform a search using the created pipeline:
```

    ]
  }
}
```

### Example: GENAI Use Case

The following example shows you how to configure an `ml_inference` search response processor with a genai model and mapping the model response to the search extension.

Step 0: Host a model
The pre-requisite is a registered genai model in opensearch. For more information about externally hosted models, see [Connecting to externally hosted models]({{site.url}}{{site.baseurl}}/ml-commons-plugin/remote-models/index/). Here is a sample predict response using a registered model, which requires a prompt and a context field.

```json
POST /_plugins/_ml/models/EOF6wJIBtDGAJRTD4kNg/_predict  
{
  "parameters": {
    "prompt":"\n\nHuman: You are a professional data analysist. You will always answer question based on the given context first. If the answer is not directly shown in the context, you will analyze the data and find the answer. If you don't know the answer, just say I don't know. Context: ${parameters.context.toString()}. \n\n Human: please summarize the documents \n\n Assistant:",

>     "context":"Dr. Eric Goldberg is a fantastic doctor who has correctly diagnosed every issue that my wife and I have had. Unlike many of my past doctors, Dr. Goldberg is very accessible and we have been able to schedule appointments with him and his staff very quickly. We are happy to have him in the neighborhood and look forward to being his patients for many years to come."

Reviewer 2: minor: Can we tune the example to make it align with the prompt ? The prompt is for data analysis , this example is not.

Author: ok, I will use this cost calculation example instead

```
  "parameters": {
     "prompt":"\n\nHuman: You are a professional data analysist. You will always answer question: Which month had the lowest customer acquisition cost per new customer? based on the given context first. If the answer is not directly shown in the context, you will analyze the data and find the answer. If you don't know the answer, just say I don't know. Context: ${parameters.context.toString()}. \n\n Assistant:",
    "context":"Customer acquisition cost: January: $50, February: $45, March: $40. New customers: January: 500, February: 600, March: 750"
  }
```
the response will be
```
Based on the context provided:

- Customer acquisition cost in January was $50 and new customers were 500. So the cost per new customer was $50/500 = $0.10

- Customer acquisition cost in February was $45 and new customers were 600. So the cost per new customer was $45/600 = $0.075 

- Customer acquisition cost in March was $40 and new customers were 750. So the cost per new customer was $40/750 = $0.053

Therefore, the month with the lowest customer acquisition cost per new customer was March, as it had the lowest cost per customer of $0.053."""
 
```

```json

PUT /_search/pipeline/my_pipeline_request_review_llm
{
  "response_processors": [
    {
      "ml_inference": {
        "tag": "ml_inference",
        "description": "This processor is going to run llm",
        "model_id": "EOF6wJIBtDGAJRTD4kNg",
        "function_name": "REMOTE",
        "input_map": [
          {
            "context": "review"
          }
        ],
        "output_map": [
          {
            "ext.ml_inference.llm_response": "response"
          }
        ],
        "model_config": {
          "prompt": "\n\nHuman: You are a professional data analyst. You will always answer questions based on the given context first. If the answer is not directly shown in the context, you will analyze the data and find the answer. If you don't know the answer, just say I don't know. Context: ${parameters.context.toString()}. \n\n Human: please summarize the documents \n\n Assistant:"
        },
        "ignore_missing": false,
        "ignore_failure": false
      }
    }
  ]
}

```
{% include copy-curl.html %}

In this configuration:

> The model_id specifies the ID of the generative AI model.

Reviewer 2: Use bulleted list ?

The function_name is set to "REMOTE", indicating an externally hosted model.
The input_map maps the review field from the document to the context field expected by the model.
The output_map specifies that the model's response should be stored in ext.ml_inference.llm_response in the search response.
The model_config includes a prompt that instructs the model on how to process the input and generate a summary.

Step 2: Index sample documents

> Index some sample documents to test the pipeline:

Reviewer 2: How about change to bulk ? That will be easier, just one API call

Author: changed to bulk
```
POST /_bulk
{"index":{"_index":"review_string_index","_id":"1"}}
{"review":"Customer acquisition cost: January: $50, New customers: January: 500."}
{"index":{"_index":"review_string_index","_id":"2"}}
{"review":"Customer acquisition cost: February: $45, New customers: February: 600."}
{"index":{"_index":"review_string_index","_id":"3"}}
{"review":"Customer acquisition cost: March: $40, New customers: March: 750."}
```

```json
POST _plugins/_ml/models/tg5p1ZEB4iWlnHsIh2U9/_predict
{
    "query_text": "today is sunny",
    "text_docs": [
        "how are you",
        "today is sunny",
        "today is july fifth",
        "it is winter"
    ]
}
```

The model returns similarity scores for each input document:

```json
{
  "inference_results": [
    {
      "output": [
        {
          "name": "similarity",
          "data_type": "FLOAT32",
          "shape": [1],
          "data": [-11.055183]
        }
      ]
    },
    {
      "output": [
        {
          "name": "similarity",
          "data_type": "FLOAT32",
          "shape": [1],
          "data": [8.969885]
        }
      ]
    },
    {
      "output": [
        {
          "name": "similarity",
          "data_type": "FLOAT32",
          "shape": [1],
          "data": [-5.736347]
        }
      ]
    },
    {
      "output": [
        {
          "name": "similarity",
          "data_type": "FLOAT32",
          "shape": [1],
          "data": [-10.0452175]
        }
      ]
    }
  ]
}
```
{% include copy-curl.html %}

Step 1: Index sample documents

> Create an index and add some sample documents:

Reviewer 2: Change to bulk?

Author: changed to bulk

```json
PUT /demo-index-0/_doc/1
{
  "dairy": "how are you"
}

PUT /demo-index-0/_doc/2
{
  "dairy": "today is sunny"
}

PUT /demo-index-0/_doc/3
{
  "dairy": "today is july fifth"
}

PUT /demo-index-0/_doc/4
{
  "dairy": "it is winter"
}

```
{% include copy-curl.html %}

Step 2: Create a search pipeline

Create a search pipeline that uses the text similarity model:

```json

> PUT /_search/pipeline/my_pipeline

Reviewer 2: Don't see `rerank` processor in this pipeline. So ml_inference processor itself support reranking?

Author: nope, I can add rerank processor to this pipeline too

{
  "response_processors": [
    {
      "ml_inference": {
        "tag": "ml_inference",
        "description": "This processor runs ml inference during search response",
        "model_id": "tg5p1ZEB4iWlnHsIh2U9",

>         "model_input": "{ \"text_docs\": ${input_map.text_docs}, \"query_text\": \"${input_map.query_text}\" }",

Reviewer 2: not sure if I misremember , have we changed `input_map` to `parameters` or not ?

Author: it's both supported

        "function_name": "TEXT_SIMILARITY",
        "input_map": [
          {
            "text_docs": "dairy",
            "query_text": "$.query.term.dairy.value"          
          }
        ],
        "output_map": [
          {
            "rank_score": "$.inference_results[*].output[*].data"
          }
        ],

>         "full_response_path": false,

Reviewer 2: `"$.inference_results[*].output[*].data"` this looks like a fulll response path, should we use `false` here ?

Author: I changed the use case to use full_response_path true

        "model_config": {},
        "ignore_missing": false,
        "ignore_failure": false
      }
    }
  ]
}
```
{% include copy-curl.html %}

This pipeline configuration:

Uses the text similarity model with ID tg5p1ZEB4iWlnHsIh2U9.
Maps the dairy field from documents to the text_docs input of the model.
Extracts the query text from the search request's term query.
Maps the model's output to a new rank_score field in the search results.

Step 3: Run the pipeline

Now, perform a search using the created pipeline:

```json
GET /demo-index-0/_search?search_pipeline=my_pipeline
{
  "query": {
    "term": {
      "dairy": {
        "value": "today"
      }
    }
  }
}
```
{% include copy-curl.html %}

> The response includes the original documents along with their calculated rank scores:

Reviewer 2: Why we only return two documents in search response ?

}
```

> ### Example: Externally hosted model

Author: there is a Externally hosted model example at line 99 already. maybe we should come up with other title? @kolchfa-aws  https://github.com/opensearch-project/documentation-website/blob/faf070f13d52f6a56cb71ea31d05f9229044bc7a/_search-plugins/search-pipelines/ml-inference-search-response.md?plain=1#L99

Reviewer 3: Updated

  }
}
```

### Example: Reranking search results using a text similarity model

The following example shows you how to configure an `ml_inference` search response processor with a text similarity model.

**Prerequisite**

> You must configure an externally hosted LLM for this use case. For more information about externally hosted models, see [Connecting to externally hosted models]({{site.url}}{{site.baseurl}}/ml-commons-plugin/remote-models/index/). Once you register the LLM, you can use the following request to test it. This request requires providing a `text` and `text_pair` fields within the `inputs` field:

Author: @kolchfa-aws  this rerank use case is using a registered text similarity model not llm model
